Zend/zend_compile.c

```diff
@@ -6739,12 +6739,94 @@ void zend_compile_class_decl(znode *result, zend_ast *ast, zend_bool toplevel) /
 			zend_error_noreturn(E_COMPILE_ERROR, "Clone method %s::%s() cannot be static",
 				ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
 		} else if (ce->clone->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+			if (ZEND_TYPE_FULL_MASK(ce->clone->common.arg_info[-1].type) != MAY_BE_VOID) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Method %s::%s() must return void",
+					ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
+			}
+		}
+	}
+
+	if (ce->__set && ce->__set->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		if (ZEND_TYPE_FULL_MASK(ce->__set->common.arg_info[-1].type) != MAY_BE_VOID) {
 			zend_error_noreturn(E_COMPILE_ERROR,
-				"Clone method %s::%s() cannot declare a return type",
-				ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
+				"Setter %s::%s() must return void",
+				ZSTR_VAL(ce->name), ZSTR_VAL(ce->__set->common.function_name));
+		}
+	}
+
+	if (ce->__unset && ce->__unset->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		if (ZEND_TYPE_FULL_MASK(ce->__unset->common.arg_info[-1].type) != MAY_BE_VOID) {
+			zend_error_noreturn(E_COMPILE_ERROR,
+				"Method %s::%s() must return void",
+				ZSTR_VAL(ce->name), ZSTR_VAL(ce->__unset->common.function_name));
 		}
 	}
 
+	if (ce->__isset && ce->__isset->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		 if (ZEND_TYPE_FULL_MASK(ce->__isset->common.arg_info[-1].type) != MAY_BE_BOOL) {
+			zend_error_noreturn(E_COMPILE_ERROR,
+				 "Method %s::%s() must return a boolean value",
+				 ZSTR_VAL(ce->name), ZSTR_VAL(ce->__isset->common.function_name));
+		 }
+	}
+
+    if (ce->__debugInfo && ce->__debugInfo->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+        if (ZEND_TYPE_FULL_MASK(ce->__debugInfo->common.arg_info[-1].type) != MAY_BE_ARRAY) {
+            zend_error_noreturn(E_COMPILE_ERROR,
+                "Method %s::%s() must return an array value",
```

Comment:
I've just checked, and `__debugInfo()` actually allows a `?array` return **type** (https://github.com/php/php-src/blob/e2f15862255701f09132e004dcc22a5f92cf402c/Zend/zend_object_handlers.c#L166). Therefore, an error message of `%s::%s() can only declare ?array as return type` or something similar (hopefully better) would be more correct.

Also, it seems to me that it's trivial to validate the return value as well (only needs an `else` with an exception thrown).

Reply:
> I've just checked, and __debugInfo() actually allows a ?array return type (

Oh shit, seriously? We need to fix that in the documentation as well then.

> Also, it seems to me that it's trivial to validate the return value as well (only needs an else with an exception thrown).

Yeah, with `zend_call_method_with_0_params`, right? But this RFC is only taking care of the method's signature. Checking for the actual return type sound like a bigger BC that I don't want to tackle right now :)

Reply:
Btw, if you want to open a PR fixing

https://github.com/php/php-src/blob/e2f15862255701f09132e004dcc22a5f92cf402c/Zend/zend_object_handlers.c#L172

feel free, I'll fix this RFC meanwhile.